src/compiler/transformers/destructuring.ts

```diff
@@ -413,17 +419,28 @@ namespace ts {
 
         /** Given value: o, propName: p, pattern: { a, b, ...p } from the original statement
          * `{ a, b, ...p } = o`, create `p = __rest(o, ["a", "b"]);`*/
-        function createRestCall<T extends Node>(value: Expression, elements: T[], getPropertyName: (element: T) => PropertyName, location: TextRange): Expression {
-            const propertyNames: LiteralExpression[] = [];
+        function createRestCall<T extends Node>(value: Expression, elements: T[], getPropertyName: (element: T) => PropertyName, location: TextRange, computedTempVariables: Expression[]): Expression {
+            const propertyNames: Expression[] = [];
             for (let i = 0; i < elements.length - 1; i++) {
-                if (isOmittedExpression(elements[i])) {
+                const element = elements[i];
+                if (isOmittedExpression(element)) {
                     continue;
                 }
-                const str = <StringLiteral>createSynthesizedNode(SyntaxKind.StringLiteral);
-                str.pos = location.pos;
-                str.end = location.end;
-                str.text = getTextOfPropertyName(getPropertyName(elements[i]));
-                propertyNames.push(str);
+                if (isComputedPropertyName(getPropertyName(element))) {
+                    // get the temp name and put that in there instead, like `_tmp + ""`
+                    const stringifiedTemp = <StringLiteral>createSynthesizedNode(SyntaxKind.StringLiteral);
+                    stringifiedTemp.pos = location.pos;
+                    stringifiedTemp.end = location.end;
+                    stringifiedTemp.text = "";
+                    propertyNames.push(createBinary(computedTempVariables.shift(), SyntaxKind.PlusToken, stringifiedTemp));

```

Comment:
That would crash with `"number"` / `"undefined"` / `null` instead :(

A workaround could be to use `String()` instead, as that will work with anything (and hopefully nobody _actually_ has a string key like that), but it would prevent easily improving `__rest` into being able to preserve (and filter) `Symbol`...

The most correct option would be `typeof x == "symbol" ? x : String(x)`.